### ISSUE TITLE
Issue adding to list after bulk add

### ISSUE THREAD

**Issue by sadatmalik** [talentbeyondboundaries/tbbtalentv2#1475]
From Stephanie:

Hello <@U055YR1QT0A>, thank you for your hard work on the great new updates.
Earlier today, I created a list for myself (not linked to SF), and when I initially selected and added candidates from my original search, it worked perfectly. However, now when I try to add more candidates, whether from other lists or manually, it seems it's not working—the original number of candidates remains unchanged.

[Slack Message](https://refugeejobsmarket.slack.com/archives/C42S4AJ73/p1728659546607299?thread_ts=1728659546.607299&cid=C42S4AJ73)

**Comment by sadatmalik** [comment]
Stephanie reported that she created a new list today (10423) and then copied some candidates to that list as selections from submission list 4571.

After copying the selected candidates, she could see those candidates in her list, but was no longer able to add any further candidates - neither individually nor via selection copy from another list.

However, in the database, I could see that additional candidates had in fact been added to her list (10423) but were not visible in the admin portal.

![Screenshot 2024-10-11 at 18 04 25](https://github.com/user-attachments/assets/260596aa-aaff-480c-9956-0467847609da)

![Screenshot 2024-10-11 at 18 04 40](https://github.com/user-attachments/assets/3ae558f7-1f91-4afe-8d2d-3c3990005c1e)

Further investigation showed that GetSavedListCandidateQuery filters out candidates who don't have opportunity for a job if the list is associated with an SF job through the presence of a job_id. I wondered if this could be causing the issue. In the database I can see that the list did in fact have a job_id:

![Screenshot 2024-10-11 at 18 03 31](https://github.com/user-attachments/assets/8dd90b40-a3db-40ef-b9b8-425313bd292e)

I tested locally to see if adding a job_id to a normal list in my local database would prevent candidates from being returned by GetSavedListCandidateQuery, which proved to be the case. I then removed the job_id from the production list, and having done so, I was able to see all the candidates that Stephanie had been adding to the list; and also a couple of candidates that I added for testing. My additions have now been manually removed.

![Screenshot 2024-10-11 at 18 06 06](https://github.com/user-attachments/assets/8d1f4622-86cf-458b-856c-5df4cc8b8d74)

![Screenshot 2024-10-11 at 18 06 25](https://github.com/user-attachments/assets/0d55ad49-e4bf-4ab8-a9a5-75201fc3c0b3)

I note that Stage is being displayed for some candidates (the ones she select copied from the submission list). But I don't think this makes sense in the context of a normal list. Neither does the show closed opps checkbox.

Removing the job_id from the list resolved the issue but I have not yet been able to identify how the job_id came to be set on this newly created list.

**Comment by sadatmalik** [comment]
Stephanie reported that she created a new list today (10423) and then copied some candidates to that list as selections from submission list 4571.

After copying the selected candidates, she could see those candidates in her list, but was no longer able to add any further candidates - neither individually nor via selection copy from another list.

However, in the database, I could see that additional candidates had in fact been added to her list (10423) but were not visible in the admin portal.

![Screenshot 2024-10-11 at 18 04 25](https://github.com/user-attachments/assets/260596aa-aaff-480c-9956-0467847609da)

![Screenshot 2024-10-11 at 18 04 40](https://github.com/user-attachments/assets/3ae558f7-1f91-4afe-8d2d-3c3990005c1e)

Further investigation showed that GetSavedListCandidateQuery filters out candidates who don't have opportunity for a job if the list is associated with an SF job through the presence of a job_id. I wondered if this could be causing the issue. In the database I can see that the list did in fact have a job_id:

![Screenshot 2024-10-11 at 18 03 31](https://github.com/user-attachments/assets/8dd90b40-a3db-40ef-b9b8-425313bd292e)

I tested locally to see if adding a job_id to a normal list in my local database would prevent candidates from being returned by GetSavedListCandidateQuery, which proved to be the case. I then removed the job_id from the production list, and having done so, I was able to see all the candidates that Stephanie had been adding to the list; and also a couple of candidates that I added for testing. My additions have now been manually removed.

![Screenshot 2024-10-11 at 18 06 06](https://github.com/user-attachments/assets/8d1f4622-86cf-458b-856c-5df4cc8b8d74)

![Screenshot 2024-10-11 at 18 06 25](https://github.com/user-attachments/assets/0d55ad49-e4bf-4ab8-a9a5-75201fc3c0b3)

I note that Stage is being displayed for some candidates (the ones she select copied from the submission list). But I don't think this makes sense in the context of a normal list. Neither does the show closed opps checkbox.

Removing the job_id from the list resolved the issue but I have not yet been able to identify how the job_id came to be set on this newly created list.

**Comment by sadatmalik** [comment]
Stephanie reported that after copying selected candidates from a search to a new list today (10423), she could see those candidates in her list, but was no longer able to add any further candidates - neither individually nor via selection copy from another list.

In the database, the additional candidates were added (10423) but not visible in the admin portal:

![Screenshot 2024-10-11 at 18 04 25](https://github.com/user-attachments/assets/260596aa-aaff-480c-9956-0467847609da)

![Screenshot 2024-10-11 at 18 04 40](https://github.com/user-attachments/assets/8f8d60ba-7565-4df1-acb4-b04002140aa5)

GetSavedListCandidateQuery filters out candidates who don't have opportunities for a job if a list is associated with an SF job through the presence of a job_id. 

A database check confirmed that the list has a job_id set:

![Screenshot 2024-10-11 at 18 03 31](https://github.com/user-attachments/assets/8dd90b40-a3db-40ef-b9b8-425313bd292e)

Removing the job_id from the list in the DB, and all candidates appear in the admin portal:

![Screenshot 2024-10-11 at 18 06 06](https://github.com/user-attachments/assets/8d1f4622-86cf-458b-856c-5df4cc8b8d74)

![Screenshot 2024-10-11 at 18 06 25](https://github.com/user-attachments/assets/645ce59a-1cc1-48f6-9c01-55340f0f4dd2)

Note that Stage is being displayed for some candidates. But I don't think this makes sense in the context of a normal list. Neither does the show closed opps checkbox.

**Conclusion:**

Removing the job_id from the list in the DB resolved the issue but we need to identify how the job_id was set on this new list.

**Comment by camerojo** [comment]
> Earlier today, I created a list for myself (not linked to SF), and when I initially selected and added candidates from my original search, it worked perfectly.

The problem is, as you say, that she thought she created a list "not linked to SF". But in fact it was linked to SF (through the job id).

That is the bug. We need to find out how she created the list.

In the latest release, the only lists associated with jobs are submission lists. We removed the capability of a user being able to associating a job with a list (which they used to be able to do). Somehow she has managed to create a list with a job. 

I suspect that this is my bug. When I removed the ability for users to be able to associate jobs with lists, I missed something.

@sadatmalik

**Comment by sadatmalik** [comment]
I did some further analysis on the existing lists in the DB to check if this may be an issue for any other lists.

I'm assuming that all list names ending in '*' are submission lists so I have ignored these from the analysis. For the remaining lists; 7 have registered_job = true. These 7 all have job_ids. They are displayed as submission lists in the admin portal:

```
select job_id,registered_job,* from saved_list
where name not like '%*'
  and registered_job is true;
```

Lists with registered_job = false but that have a job_id set are 479 in number:

```
select job_id,registered_job,* from saved_list
where name not like '%*'
  and registered_job is false
  and job_id is not null
order by id desc;
```

Checking one of these lists (10367) I can add a test candidate from the admin portal, and while it is correctly updated in the DB, it's not returned by GetSavedListCandidateQuery. I'm not fully sure of the implications of setting job_id to null for these lists; but without doing so the lists will not behave as expected.

@camerojo

**Comment by camerojo** [comment]
Yes - I chose to leave existing lists with jobs alone. Maybe that was a mistake - but that is what I did. The rationale was not to mess with lists that people already had associated with jobs. I think it probably was a mistake because those lists no longer operate as previously - adding a candidate to those lists no longer automatically creates a candidate opp on SF. But it does not make sense for a candidate to be in a list associated with a job when that candidate is not also associated with the job. Hence the anomaly. 

We need to think some more about how to handle those legacy lists with jobs associated. Maybe just removing the job id is the best way forward - but let's think about that a bit more. Although there are 479 such lists - some are very old and the number of users involved is not huge (a lot belong to me, you and Caroline)

What is the mystery here is she says that she created a new list. I don't understand how the job got associated.

**Comment by camerojo** [comment]
@sadatmalik

**Comment by camerojo** [comment]
OK - problem solved. My bug. Searches can also have jobs associated with them. When you save a selection from a search to a list, the job associated with the search is also associated with the list.

That needs to change now. In fact I think we can drop it altogether. It made some sense before when all lists and searches could be associated with jobs. It does not now.

The relevant code is in CandidateSavedListServiceImpl.copy

<pre>

        //Set any specified Job Opportunity
        final Long jobId = request.getJobId();
        if (jobId != null) {
            targetList.setSfJobOpp(salesforceJobOppService.getJobOpp(jobId));
        }
</pre> 

@sadatmalik 

She created a list - not associated with any job.

But then she copied a selection from a search which was associated with a job - and that job got copied across to the list.

**Comment by sadatmalik** [comment]
Yes, that's it - nice one @camerojo !

When copying selected candidates from a search associated with a job (as a suggested search) to a list, the job's job_id from the suggested search is assigned to the list.

I'll push a Hotfix to master and attach a PR for staging to this issue.
